engine/src/main/java/org/terasology/config/flexible/SettingImpl.java

```diff
@@ -0,0 +1,174 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.config.flexible;
+
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.config.flexible.validators.SettingValueValidator;
+import org.terasology.engine.SimpleUri;
+
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.util.List;
+
+/**
+ * {@inheritDoc}
+ * @param <T> The type of the value this {@link SettingImpl} contains.
+ */
+public class SettingImpl<T> implements Setting<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SettingImpl.class);
+    private final T defaultValue;
+
+    private SimpleUri id;
+
+    private T value;
+
+    private String humanReadableName;
+
+    private String description;
+    private SettingValueValidator<T> validator;
+    private List<PropertyChangeListener> subscribers;
+
+    /**
+     * Creates a new {@link SettingImpl} that does not use a validator with the given id and default value.
+     * @param id the id of the setting.
+     * @param defaultValue the default value of the setting.
+     */
+    public SettingImpl(SimpleUri id, T defaultValue) {
+        this(id, defaultValue, null);
+    }
+
+    /**
+     * Creates a new {@link SettingImpl} with the given id, default value and validator.
+     * @param id the id of the setting.
+     * @param defaultValue the default value of the setting.
+     * @param validator the validator to be used to validate values.
+     */
+    public SettingImpl(SimpleUri id, T defaultValue, SettingValueValidator<T> validator) {
+        this.id = id;
+
+        this.validator = validator;
+
+        if (!validate(defaultValue))
+            throw new IllegalArgumentException("The default value must be a valid value.");
+
+        this.defaultValue = defaultValue;
+        this.value = this.defaultValue;
+
+        this.subscribers = null;
+    }
+
+    private void dispatchChangedEvent(PropertyChangeEvent event) {
+        for (PropertyChangeListener subscriber : subscribers) {
+            subscriber.propertyChange(event);
+        }
+    }
+
+    private boolean validate(T value) {
+        return validator == null || validator.validate(value);
+    }
+
+    /**
+     * Subscribe a listener that will be notified when the value stored in the setting changes.
+     */
+    public void subscribe(PropertyChangeListener listener) {
+        if (subscribers == null) {
+            subscribers = Lists.newArrayList();
+        }
+
+        subscribers.add(listener);
```

Comment:
Must have sneaked under the radar. This will require `Setting` to not implement `GeneralSubscribable`, though. If that's okay, I can put the `subscribe` and `unsubscribe` methods in the `Setting` interface itself.